index.js

```diff
@@ -16,117 +16,123 @@ function helmetPlugin (fastify, options, next) {
+      : configuration
 
-    const cspMiddleware = helmet.contentSecurityPolicy({ directives, reportOnly: cspReportOnly })
-    cspMiddleware(request.raw, reply.raw, next)
+    return helmet(helmetConfiguration)(request.raw, reply.raw, done)
   }
 }
 
-function buildRouteHooks (configuration, routeOptions, decorateOnly) {
-  if (Array.isArray(routeOptions.onRequest)) {
-    routeOptions.onRequest.push(addHelmetReplyDecorator)
-  } else if (typeof routeOptions.onRequest === 'function') {
-    routeOptions.onRequest = [routeOptions.onRequest, addHelmetReplyDecorator]
-  } else {
-    routeOptions.onRequest = [addHelmetReplyDecorator]
-  }
-
-  const middleware = helmet(configuration)
+async function buildHelmetOnRoutes (request, reply, configuration, enableCSP) {
+  if (enableCSP === true) {
+    const cspDirectives = configuration.contentSecurityPolicy
+      ? configuration.contentSecurityPolicy.directives
+      : helmet.contentSecurityPolicy.getDefaultDirectives()
+    const cspReportOnly = configuration.contentSecurityPolicy
+      ? configuration.contentSecurityPolicy.reportOnly
+      : undefined
 
-  function addHelmetReplyDecorator (request, reply, next) {
-    // We decorate `reply.helmet` with all helmet middleware functions
-    // NB: we allow users to pass a custom helmet options object with a fallback
-    // to global helmet configuration.
-    reply.helmet = (opts) => opts
-      ? helmet(opts)(request.raw, reply.raw)
-      : helmet(configuration)(request.raw, reply.raw)
+    // We get the csp nonce from the reply
+    const { script: scriptCSPNonce, style: styleCSPNonce } = reply.cspNonce
 
-    next()
-  }
+    // We prevent object reference: https://github.com/fastify/fastify-helmet/issues/118
+    const directives = { ...cspDirectives }
 
-  if (decorateOnly) {
-    return
-  }
+    // We push nonce to csp
+    // We allow both 'script-src' or 'scriptSrc' syntax
+    const scriptKey = Array.isArray(directives['script-src']) ? 'script-src' : 'scriptSrc'
+    directives[scriptKey] = Array.isArray(directives[scriptKey]) ? [...directives[scriptKey]] : []
+    directives[scriptKey].push(`'nonce-${scriptCSPNonce}'`)
+    // allow both style-src or styleSrc syntax
+    const styleKey = Array.isArray(directives['style-src']) ? 'style-src' : 'styleSrc'
+    directives[styleKey] = Array.isArray(directives[styleKey]) ? [...directives[styleKey]] : []
+    directives[styleKey].push(`'nonce-${styleCSPNonce}'`)
 
-  // At this point `routeOptions.onRequest` is an array
-  // we just have to push our `onRequest` function
-  routeOptions.onRequest.push(onRequest)
+    const contentSecurityPolicy = { directives, reportOnly: cspReportOnly }
+    const mergedHelmetConfiguration = Object.assign(Object.create(null), configuration, { contentSecurityPolicy })
 
-  function onRequest (request, reply, next) {
-    middleware(request.raw, reply.raw, next)
+    helmet(mergedHelmetConfiguration)(request.raw, reply.raw, done)
```

Comment:
I do not think it is the correct approach for the done method.
The hook may end before throwing error.

The done method should inherit from the onRequest hook, there are problem in previous version because it do not terminate the the script in the correct position and it run the done method twice.

PS. I put the previous comment in wrong place.

Reply:
One more question, will the error be catched inside a callback hook?

Reply:
> One more question, will the error be catched inside a callback hook?

In the current code it correctly forwards helmet's internal errors (I added a test to make sure it was the case)
and fastify correctly catch the error if we use an `onError` hook. (I can add a test to improve the number of use cases covered)